math/minuit2/src/Minuit2Minimizer.cxx

```diff
@@ -421,11 +426,11 @@ bool Minuit2Minimizer::Minimize() {
    }
 
    // internal minuit messages
-   MnPrint::SetLevel(printLevel );
    fMinimizer->Builder().SetPrintLevel(printLevel);
 
    // switch off Minuit2 printing
-   int prev_level = (printLevel <= 0 ) ?   TurnOffPrintInfoLevel() : -2;
+   const int prev_level = (printLevel <= 0 ) ?   TurnOffPrintInfoLevel() : -2;
+   const int prevGlobalLevel = MnPrint::SetGlobalLevel(printLevel);
```

Comment:
Here I think we need to adapt the printLevel of MInimizer to the unse used in MnPrint
Probably we should do something like: 
Minuit2Minimizer::PrintLevel() < 0  --> all message suppressed MnPrint default level = 0
                                                = 0   ->  Only Error message printed  (MnPrint::SetGlobalLevel(1)
                                                = 1   ->  Error + warning message printed: MnPrint::SetGlobalLevel(2)
                                                = 2,3   -> Info message printed : MnPrint::SetGlobalLevel(3)
                                                > 3       -> Debug mode  : MnPrint::SetGlobalLevel(4)

This adaptation is needed to avoid changing output messages drammatically running workflow using MInuit2 with this MinimizerInterface from ROOT (e.g. ROOT fitting, Roofit/RooStats , etc....).
I could eventually apply also after this PR is merged with a following PR

Reply:
Thank you Hans for this nice improvement in the Minuit2 logging!

I see in Debug mode that a lots of messages are printed. Is it possible to filter the messages on the class that generates them ? That would be useful

Another useful thing would be maybe to have two level of debug verbosity. For example when doing a fit with many parameters I guess you don;t want to print all parameter values and matrix values, because the messages will be unreadable. Maybe in that case only function values could be printed (and all in one line).
Anyway these improvements could also be applied afterwards with an additional PR. 

Cheers

Reply:
> I see in Debug mode that a lots of messages are printed. Is it possible to filter the messages on the class that generates them ? That would be useful

I agree that this is useful and on my to-do list for follow-up work. There are several options, please let me know what you think about them:
1) Make all prints one-line, so that one can grep out those of interest, just using the class name.
Pros:
- Simple to implement, simple to understand and simple to use
Cons:
- Lines will become very long in some cases
- Printing vectors and matrices in one line may hurt readability a lot
2) Add a `SetPrintLevel` method to each class, so that the print level can be set individually for each class. If no specific level is chosen, use the global default level.
Pros:
- Very flexible. Can use multi-line printouts.
- Seems to be a standard solution. I have seen such an implementation e.g. in the LHCb software framework.
- Easy to pass down print level recursively, e.g. if MnMigrad is calling MnLineSearch, MnLineSearch can use the print level of MnMigrad.
Cons (sort of):
- Some interface needs to be added in ROOT to allow setting the debug level individually for each class. This needs some thought. In iminuit it is not an issue, because I use the Minuit2 classes directly, but ROOT has this abstract Minimizer interface on top of several minimizers, so it is not clear how to pass the information down to Minuit2. This could perhaps be done with a string based interface: `SetPrintLevel("MnLineSearch", 3);` String-based interfaces are not ideal, but they have the advantage of being perfectly decoupled from the actual implementation below. An alternative is to use an enum instead of a string, but that exposes some details of the Minuit2 class structure.
3) Build a global hash-map into MnPrint which stores individual print levels for each class using a mapping from class name (string) to print level.
Pros:
- A comparably simple extension of the current design.
- All control of print-levels still goes through MnPrint, one does not have to set print-levels individually on each class with a method on that class.
Cons:
- String-based interfaces are brittle, typos can only be detected at run-time.
- Like in 2) some interface in ROOT is needed to allow users to set the print level.
- Passing-down print levels is not so straight forward in this design.

I suppose the best short-term solution is 3). We should decide whether we generally want classes that are internally called by other classes to inherit the print level of the caller. If that's desired, then solution 2) is cleaner. If not, then I would go for 3). I cannot predict at this point which design will be better in practice.

If we decide to go for 3), should I make the changes in this PR or implement that in a follow up PR?

In general, I am also very interested in more fine-grained control on the printing also for iminuit, although I would also like to keep an option to "show me everything", because it really helped me a lot to understand what Migrad actually does and in what order.

Reply:
Hi Hans, 
Thank you very much for your detailed analysis on how to implement a fine grained printing in Minuit2. 
My preference would be probably to go for option (3) that is maybe less clean, but also more practical to implement. 
However I see that option (2) has the advantage to have a recursivw print level. 

Being practical,  we would not need a detailed level for every single class, but for maybe some group of class implementing a specific algorithm. For example you would like to debug the iterations in the VariableMetric algorithm (MnMigrad) or the computation of numerical derivatives or the line search algorithm. 
This could be controlled maybe easier with a global hash map as you suggest.
There is often the use case you want to debug a lower level class (e.g. MnLineSearch) but not debugging an higher level (MnMigrad) or the opposite (e.g. debugging MnMinos but not MnMigrad). 
Of course there is also the case you want to see everything. 
A simpler solution is also having multiple debug levels which apply to all Minuit. 

Anyway, I agree it is better to merge first this PR and do improvements eventually in a following one.  

For this PR, I would add just the fix for the Info message in MnSeedGenerator. I can then commit later a subsequent PR to adapt the Minuit2Minimizer setting of the Print Level. 

Thank you !

Reply:
Ok, will do! Thank you for the discussion! What you said gave me some food for thought.

I think we both prefer option 3) then, but this approach should still make it easy to distinguish, staying with your example, MnLineSearch called by MnMigrad from MnLineSearch called by MnMinos. This means that the MnPrint instances need to know when they are nested.  I think this can be achieved with a minor modification of the current design at the cost of again manipulating global state.

Since MnPrint is already a scope guard, it could modify a global prefix. Pseudo-code:
```c++
{
  // in MnMigrad scope
  MnPrint print("MnMigrad"); // adds MnMigrad to global prefix
 
  print.Info("foo"); // prints MnPrint: foo
  // ... at some point an instance of MnLineSearch is created
  {
     // in MnLineSearch scope
     MnPrint print("MnLineSearch"); // adds MnLineSearch to global prefix
     // ...
     print.Info("bar"); // print MnPrint::MnLineSearch: bar
     // end of scope, remove MnLineSearch from global prefix
  }
  // end of scope, remove MnMigrad from global prefix
}
```
The only issue is to make this thread-safe. The global prefix must be a thread-local variable.

Reply:
In https://github.com/root-project/root/pull/6705/commits/a6968f4a79a3728385d1461400920e03d829e569, I implemented this idea and I made the global variables related to MnPrint thread_local (which I should have done anyway).

I like these stacking prefixes, because they show how algorithm calls are nested. The downside is that the prefix gets rather long (which I find acceptable). An example from iminuit:
```
I MnApplication:MnSeedGenerator: Initial state: FCN =                 0 Edm =                 0 NCalls =      9
I MnApplication:ModularFunctionMinimizer:VariableMetricBuilder: Start iterating until Edm is < 0.0002 with call limit = 420
I MnApplication:ModularFunctionMinimizer:VariableMetricBuilder:    0 - FCN =                 0 Edm =                 0 NCalls =      9
W MnApplication:ModularFunctionMinimizer:VariableMetricBuilder:MnHesse:MnPosDef: Matrix forced pos-def by adding to diagonal 0.002
I MnApplication:ModularFunctionMinimizer:VariableMetricBuilder: After Hessian
I MnApplication:ModularFunctionMinimizer:VariableMetricBuilder:    1 - FCN =                 0 Edm =                 0 NCalls =     19
```